### DISCUSSION TITLE
how do I get the raw logits from the YOLOv11 model prediction?

### DISCUSSION THREAD

**Discussion by ambient-udeshmukh** [ultralytics/ultralytics#19767] [General]
Hello.
I have a yolov11 model which I have converted to onnx using the `model.export()` command.
In order to check the outputs, when I run the model in torch, I get a list of `ultralytics.engine.results.Result` objects, whereas when I run the onnx model, I get raw logits of shape 1 x 33 x 8400

I want to get the raw logits from the torch model to make sure that the onnx model is generating the same output as the torch model.

Does anybody know how to access the raw logits (before the post processing step)?

Best Regards

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ambient-udeshmukh, thank you for your question and for using Ultralytics 🚀! 

For your inquiry about accessing raw logits from the YOLO11 model, it seems that you are comparing outputs from the PyTorch and ONNX versions of the model. To assist further:

- If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it effectively.  
- If this is a ❓ Question, consider sharing additional details like the specific code snippet generating the `ultralytics.engine.results.Result` objects and any ONNX runtime setup you are using. This will help us better understand your setup and guide you.

We also recommend checking out the [Docs](https://docs.ultralytics.com/) for detailed information on model outputs and post-processing steps. You may find the Python-specific [usage examples](https://docs.ultralytics.com/usage/python/) particularly helpful for working directly with YOLO models.

## Upgrade

To ensure you are working with the latest features and fixes, upgrade your `ultralytics` package in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

For experimentation, YOLO models can be run in verified environments pre-configured with all dependencies:

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community

For additional support, join the Ultralytics community:
- [Discord](https://discord.com/invite/ultralytics) for real-time chat 🎧
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to share insights with others

An Ultralytics engineer will review your discussion and assist you soon. Thank you for your patience and for being part of the Ultralytics community! 😊

**Comment by glenn-jocher** [comment]
@ambient-udeshmukh to access raw YOLO11 model outputs before post-processing in PyTorch, use the model directly on preprocessed images:

```python
from ultralytics import YOLO

model = YOLO('yolo11n.pt')
img = model.transforms('path/to/image.jpg')  # preprocess
img = img.unsqueeze(0).to(model.device)  # add batch dim
raw_outputs = model.model(img)  # raw logits tensor
```

This matches the ONNX output shape and bypasses post-processing. For detailed preprocessing steps, see the [Python predict documentation](https://docs.ultralytics.com/usage/python/#predict).

**Reply by ambient-udeshmukh** [reply]
ohh i see.. instead of running `torch_model(dummy_input)`
I should be running `torch_model.model(dummy_input)`

Thank you for the help @glenn-jocher

**Reply by glenn-jocher** [reply]
@ambient-udeshmukh you're welcome! For anyone else needing to access raw model outputs before post-processing in PyTorch, remember to use `model.model()` directly on properly preprocessed images as shown in the [Python predict documentation](https://docs.ultralytics.com/usage/python/#predict). This ensures consistency with exported ONNX outputs.

**Reply by ambient-udeshmukh** [reply]
@glenn-jocher I was able to convert the torch model to tensorRT using the export method.
Now I want to run a dummy input tensor through this model, and get the raw logits.

I dug into the source codes, and it looks like it is mandatory to go over `preprocess`, `inference`, and `postprocess` step.
Looks like currently there is no user controlled flag which allows us to run the inference **without** the postprocessing step.

This is the part of the source code that I am looking at <this happens in engine.predictor.BasePredictor>:
```
# Preprocess
  with profilers[0]:
      im = self.preprocess(im0s)

  # Inference
  with profilers[1]:
      preds = self.inference(im, *args, **kwargs)
      if self.args.embed:
          yield from [preds] if isinstance(preds, torch.Tensor) else preds  # yield embedding tensors
          continue

  # Postprocess
  with profilers[2]:
      self.results = self.postprocess(preds, im, im0s)
```

Does this means that if the user is running inference on a tensorRT model, there is no way to get the raw logits (output before the postprocessing step)?

Best Regards

**Reply by glenn-jocher** [reply]
To access raw TensorRT model outputs before post-processing, you can directly use the model's forward method with preprocessed input. Here's how:

```python
model = YOLO('yolo11n.engine')
im = model.transforms('image.jpg')  # preprocess
raw_outputs = model.model.forward(im, augment=False, visualize=False)
```

This bypasses the post-processing steps while maintaining proper input formatting. For implementation details, see the [autobackend forward method documentation](https://docs.ultralytics.com/reference/nn/autobackend/#autobackend.forward).

**Reply by ambient-udeshmukh** [reply]
Thank you @glenn-jocher 
I was comparing the run time comparison of the torch model vs onnx vs tensorRT. My intention is to recreate the gains mentioned on the ultralytics website (5x for tensorRT).
but i dont get similar numbers. My numbers look as follows:

![image](https://github.com/user-attachments/assets/8281aa9a-58b4-4952-9517-8f72cc4cf4ad)

I am using the default parameters for the onnx conversion (like this):
```
torch_model = YOLO("ultralytics_official_models/yolo11n.pt").to(device='cuda')
torch_model.export(format="onnx",
                            imgsz=(480, 640),
                            opset=19,
                            dynamic=False)
```

I convert the onnx model to TRT using this:
```
trtexec --onnx=ultralytics_official_models/yolo11n.onnx --saveEngine=ultralytics_official_models/yolo11n_fp16.trt --fp16
```

From the table, we can see that TRT-FP32 actually performs worse than the torch-FP32 model.

Any idea what I am doing wrong?

**Reply by glenn-jocher** [reply]
@ambient-udeshmukh for optimal TensorRT performance with YOLO11, use Ultralytics' native export (`model.export(format='engine')`) rather than trtexec. Our integration handles layer fusion and kernel tuning automatically. Key tips:  
1. Export directly to TensorRT with `half=True` for FP16  
2. Enable `dynamic=True` to allow batch-size optimizations  
3. Use square resolutions (640x640) for best performance  

The speed benchmarks in our docs use Ultralytics-engineered export pipelines. You can find the exact configurations in the [TensorRT export documentation](https://docs.ultralytics.com/integrations/tensorrt/#export-configurations).

**Reply by ambient-udeshmukh** [reply]
@glenn-jocher 
Here are a few experiments that I did, the way I set these experiments and my findings from these experiments:

System setup:
- I am using a AWS-EC2 (g5.2xlarge) as my compute device. This contains a nvidia A10G GPU. I am using CUDA 12.8 with NVCC V12.8.93

Experiment set up:
1) Experiment 1: Using ultralytics's export routine with `non square` image resolution:
```
torch_model.export(format="engine",
                                imgsz=(480, 640),
                                batch=SESSION_CONFIG['BATCH_SIZE'],
                                opset=19,
                                dynamic=True)

trt_model = YOLO("yolo_direct_fp32.engine", task='detect')
trt_output = trt_model.predict(dummy_image, imgsz=(480, 640), verbose=False)
```
**Observations**:
- `parity mismatch`: `trt_output` does not match the output from the torch model
- `time performance`: both these methods are taking ~2.97 seconds (compared to pytorch which takes ~3.2 seconds)
- `gpu utilization`: when the test is running the GPU utilization looks good
Test Setup:
```
trt_model = YOLO("yolo_direct_fp32.engine", task='detect')

# warmup run
for _ in range(2):
    _ = trt_model.predict(dummy_image, imgsz=(480, 640), verbose=False)

# start the test
start = time.time()
for _ in range(num_iterations):
    trt_output = trt_model(dummy_image, imgsz=(480, 640), verbose=False)
    
print(f'time taken by trt model = {time.time() - start}')
```


2) Experiment 2: Using ultralytics's export routine with `square` image resolution (640 x 640):
```
torch_model.export(format="engine",
                                imgsz=640,
                                batch=SESSION_CONFIG['BATCH_SIZE'],
                                opset=19,
                                dynamic=True)

trt_model = YOLO("yolo_direct_fp32.engine", task='detect')
trt_output = trt_model.predict(dummy_image, imgsz=640, verbose=False)
```
**Observations**:
- `parity mismatch`: `trt_output` does not match the output from the torch model
- `time performance`: both these methods are taking ~2.3 seconds (compared to pytorch which takes ~4.3 seconds)
- `gpu utilization`: when the test is running the GPU utilization looks good

Miscellaneous observations:
- Converting the model to trt using trtexec and running inference gave me parity with the torch output
- model with trt exec was ~21% slower than the model generated with ultralytics export
- 480x640 resolution was consistently faster than 640x640

**_NOTE_**: All of these comparisons was with torch-fp32 vs TRT-FP32

**Reply by ambient-udeshmukh** [reply]
## Parity test - This is the experiment to monitor the raw logits from torch vs tensorRT: 
```
trt_model = YOLO("yolo_direct_fp32.engine", task='detect')

# Warmup runs
for _ in range(2):
    with torch.no_grad():
        torch_model_det(dummy_image, verbose = False)
    trt_model(dummy_image, verbose = False)
print('warm up run completed. 🔥')

## Parity check
logit_torch = torch_model_det.model(dummy_image)
logit_trt = trt_model.predictor.inference(dummy_image)
```

Here I see that the `logit_torch` arenot matching `logit_trt`

**Reply by glenn-jocher** [reply]
For TensorRT parity testing, ensure identical preprocessing by using `model.transforms()` on your input before inference. The observed differences likely stem from preprocessing variations rather than model architecture. When comparing raw outputs, directly access the engine's output tensor after proper preprocessing as shown in the [TensorRT inference documentation](https://docs.ultralytics.com/integrations/tensorrt/#inference-with-tensorrt). For performance optimization, we recommend using square resolutions and FP16 as shown in our benchmarks.

**Reply by ambient-udeshmukh** [reply]
I have verified that the inputs look accurate. Here is a summary of my findings:
- When running the model using image_name as input, the outputs look accurate
- When reading the image offline and converting it to a tensor, and then sending it through the model, the outputs are incorrect (all the preprocess functions are verified here). I ensured this by monitoring the inputs at the line:.
`preds = self.inference(im, *args, **kwargs)`  inside of the `BasePredictor` class

my suspicion is that the `bindings` are not accessing the cuda memory correctly.

**Reply by glenn-jocher** [reply]
For TensorRT CUDA memory binding issues, ensure your input tensor matches the exact preprocessing steps and memory layout expected by the engine. The TensorRT inference documentation shows proper CUDA memory handling techniques. Verify your tensor is properly allocated on GPU with correct shape/dtype and NCHW format, and consider adding warmup iterations before benchmarking.

**Comment by ambient-udeshmukh** [comment]
I think i got all my questions answered. we can close this discussion.

**Reply by glenn-jocher** [reply]
@ambient-udeshmukh to access raw YOLO11 model outputs before postprocessing in PyTorch, you can directly call the model with preprocessed tensors:

```python
model = YOLO('yolo11n.pt')
im = model.preprocess('bus.jpg')  # preprocessed tensor
raw_output = model.model(im)      # raw model output before postprocessing
```

This `raw_output` tensor will match the ONNX model's output shape when using identical input preprocessing. For details on preprocessing see the Ultralytics YOLO Predict documentation.
